internal/store/db/redis/store.go

```diff
@@ -16,37 +16,208 @@
+		if err != nil {
+			return nil, err
+		}
+		objects = append(objects, model.ToContract())
+	}
+
+	return objects, nil
 }
 
 // Update replaces the data currently in the store with the provided data.
-func Update(o contracts.StoredObject) error {
+func (c Client) Update(o contracts.StoredObject) error {
+	err := o.ValidateContract(true)
+	if err != nil {
+		return err
+	}
+
+	conn := c.Pool.Get()
+	defer conn.Close()
+
+	err = removeFromDB(conn, o)
+	if err != nil {
+		return err
+	}
+
+	_, err = addToDB(conn, o)
+	if err != nil {
+		return err
+	}
+
 	return nil
 }
 
 // RemoveFromStore removes an object from the data store.
-func RemoveFromStore(id string) error {
-	return nil
+func (c Client) RemoveFromStore(o contracts.StoredObject) error {
+	err := o.ValidateContract(true)
+	if err != nil {
+		return err
+	}
+
+	conn := c.Pool.Get()
+	defer conn.Close()
+
+	return removeFromDB(conn, o)
 }
 
 // Disconnect ends the connection.
-func Disconnect() error {
-	return nil
+func (c Client) Disconnect() error {
+	return c.Pool.Close()
 }
 
 // NewClient provides a factory for building a StoreClient
 func NewClient(config db.DatabaseInfo) (interfaces.StoreClient, error) {
-	return nil, nil
+	once.Do(func() {
+		connectionString := fmt.Sprintf("%s:%d", config.Host, config.Port)
+		opts := []redis.DialOption{
+			redis.DialPassword(config.Password),
+			redis.DialConnectTimeout(time.Duration(config.Timeout) * time.Millisecond),
+		}
+
+		dialFunc := func() (redis.Conn, error) {
+			conn, err := redis.Dial(
+				"tcp", connectionString, opts...,
+			)
+			if err != nil {
+				return nil, fmt.Errorf("Could not dial Redis: %s", err)
+			}
+			return conn, nil
+		}
+		currClient = &Client{
+			Pool: &redis.Pool{
+				IdleTimeout: time.Duration(config.Timeout) * time.Millisecond,
+				/* The current implementation processes nested structs using concurrent connections.
+				 * With the deepest nesting level being 3, three shall be the number of maximum open
+				 * idle connections in the pool, to allow reuse.
+				 * TODO: Once we have a concurrent benchmark, this should be revisited.
+				 * TODO: Longer term, once the objects are clean of external dependencies, the use
+				 * of another serializer should make this moot.
+				 */
+				MaxIdle: config.MaxIdle,
+				Dial:    dialFunc,
+			},
+			BatchSize: config.BatchSize,
+		}
+	})
+
+	return currClient, nil
+}
+
+func addToDB(conn redis.Conn, o contracts.StoredObject) (string, error) {
+	exists, err := redis.Bool(conn.Do("EXISTS", o.ID))
+	if err != nil {
+		return "", err
+	} else if exists {
+		return "", errors.New("object exists in database")
+	}
+
+	var model models.StoredObject
+	model.FromContract(o)
+
+	json, err := model.MarshalJSON()
+	if err != nil {
+		return "", err
+	}
+
+	_ = conn.Send("MULTI")
+	// store the object's representation
+	_ = conn.Send("SET", model.ID, json)
+	// store the association with this ASK
+	_ = conn.Send("SADD", model.AppServiceKey, model.ID)
+
```

Comment:
If the AppServiceKey is used as a key anywhere else in the platform we will have challenges. I suggest adding a namespace prefix to the AppServiceKey such as the currently unused collection constant.

Reply:
@lenny-intel , should this be as simple as `redisCollection + ":" + model.AppServiceKey`?

Reply:
Yes